src/software/ai/hl/stp/play/free_kick_play_test.cpp

```diff
@@ -83,3 +96,97 @@ TEST_F(FreeKickPlayTest, test_free_kick_play_on_friendly_half)
             terminating_validation_functions, non_terminating_validation_functions,
             Duration::fromSeconds(10));
 }
+
+TEST_F(FreeKickPlayIsApplicableInvariantHoldsTest, test_is_applicable_set_ball_center)
+{
+    // isApplicable Test 1: GameState is set to isOurFreeKick and
+    // min_dist_to_corner >= cornerKickPlay:BALL_IN_CORNER_RADIUS.
+    // Set team with possession to FRIENDLY
+    world.updateGameState(::TestUtil::createGameState(RefereeCommand::DIRECT_FREE_US,
+                                                      RefereeCommand::HALT));
+
+    world = ::TestUtil::setBallPosition(world, Point(0, 0), Timestamp::fromSeconds(0));
+
+    world.setTeamWithPossession(TeamSide::FRIENDLY);
+
+    ASSERT_TRUE(free_kick_play.isApplicable(world));
+
+    world.updateGameState(::TestUtil::createGameState(RefereeCommand::INDIRECT_FREE_US,
+                                                      RefereeCommand::HALT));
+
+    ASSERT_TRUE(free_kick_play.isApplicable(world));
+}
+
+TEST_F(FreeKickPlayIsApplicableInvariantHoldsTest,
+       test_is_applicable_set_ball_enemy_corners)
+{
+    // isApplicable Test 2: GameState is set to isOurFreeKick and
+    // min_dist_to_corner < cornerKickPlay:BALL_IN_CORNER_RADIUS.
+    // Team with possession is FRIENDLY
+    world.updateGameState(::TestUtil::createGameState(RefereeCommand::DIRECT_FREE_US,
+                                                      RefereeCommand::HALT));
+
+    world = ::TestUtil::setBallPosition(world, world.field().enemyCornerPos(),
+                                        Timestamp::fromSeconds(0));
+
+    world.setTeamWithPossession(TeamSide::FRIENDLY);
+
+    ASSERT_FALSE(free_kick_play.isApplicable(world));
+
+    world = ::TestUtil::setBallPosition(world, world.field().enemyCornerNeg(),
+                                        Timestamp::fromSeconds(0));
+
+    ASSERT_FALSE(free_kick_play.isApplicable(world));
+}
+
+TEST_F(FreeKickPlayIsApplicableInvariantHoldsTest, test_is_applicable_enemy_free_kick)
+{
+    // isApplicable Test 3: GameState is set to isTheirFreeKick and
+    // min_dist_to_corner > cornerKickPlay:BALL_IN_CORNER_RADIUS.
+    // Team with possession is ENEMY
+    world.updateGameState(::TestUtil::createGameState(RefereeCommand::DIRECT_FREE_THEM,
+                                                      RefereeCommand::HALT));
+
+    world = ::TestUtil::setBallPosition(world, Point(0, 0), Timestamp::fromSeconds(0));
+
+    world.setTeamWithPossession(TeamSide::ENEMY);
+
+    ASSERT_FALSE(free_kick_play.isApplicable(world));
```

Comment:
why isn't there an invariantsHold check in this situation?

Reply:
I suggested we separate invariantHolds and isApplicable to two different tests with the same state so we aren't testing more than one thing per unit test